bc_obps/reporting/service/report_activity_save_service.py

```diff
@@ -0,0 +1,224 @@
+import uuid
+from django.db import transaction
+from registration.models.activity import Activity
+from reporting.models.activity_json_schema import ActivityJsonSchema
+from reporting.models.activity_source_type_json_schema import ActivitySourceTypeJsonSchema
+from reporting.models.facility_report import FacilityReport
+from reporting.models.fuel_type import FuelType
+from reporting.models.gas_type import GasType
+from reporting.models.report_activity import ReportActivity
+from reporting.models.report_emission import ReportEmission
+from reporting.models.report_fuel import ReportFuel
+from reporting.models.report_source_type import ReportSourceType
+from reporting.models.report_unit import ReportUnit
+from reporting.models.source_type import SourceType
+from reporting.service.utils import exclude_keys, retrieve_ids
+
+
+class ReportActivitySaveService:
+    """
+    Service that handles a json objects coming from an activity form:
+    - Splits out json data in, and saves individual parts into a ReportActivity object
+    - With a report_version, a facility and an activity, fetches the report_activity and all its dependent data and returns json data for the form
+
+    usage:
+    service = ReportActivitySaveLoadService(report_version_id, facility_id, activity_id)
+    service.save(json_data)
+
+    service.load()
+    """
+
+    facility_report: FacilityReport
+    activity: Activity
+    user_guid: uuid.UUID
+
+    def __init__(self, report_version_id: int, facility_id: uuid.UUID, activity_id: int, user_guid: uuid.UUID):
+        self.user_guid = user_guid
+        self.activity = Activity.objects.get(pk=activity_id)
+        self.facility_report = FacilityReport.objects.select_related('report_version', 'report_version__report').get(
+            report_version_id=report_version_id, facility_id=facility_id
+        )
+
+    @transaction.atomic()
```

Comment:
@Sepehr-Sobhani what's the best practice about atomic transactions? Should that be specified only at the API level? We risk nesting transactions otherwise, I don't know the performance penalty for that

Reply:
Based on my knowledge having it at the top level of the function should be enough (service layer) but we can also manage nested transactions with context manager: https://docs.djangoproject.com/en/5.1/topics/db/transactions/#:~:text=and%20as%20a%20context%20manager%3A

Note: I wasn't able to find a way to automate testing atomic transactions so I always do it manually.